runtime/src/utils/buffer/write.rs

```diff
@@ -176,153 +217,100 @@ impl<B: Blob> Blob for Write<B> {
-
-        // Update position to maintain "buffer at tip" invariant.
-        //
-        // Position should advance to the end of this write if it extends the logical blob
-        let write_end = offset + data_len as u64;
-        if write_end > inner.position {
-            inner.position = write_end;
+        // Write cannot be merged, so flush the buffer if the range overlaps, and check if merge is
+        // possible after.
+        if !buffer.is_empty() && buffer.offset < end_offset {
+            let (old_buf, buffer_offset) = buffer.take_buf();
+            self.blob.write_at(old_buf, buffer_offset).await?;
+
+            if buffer.merge(buf.as_ref(), offset) {
+                return Ok(());
+            }
         }
 
+        // Write could not be merged (exceeds buffer capacity or outside its range), so write
+        // directly. Note that we end up writing an intersecting range twice: once when the buffer
+        // is flushed above, then again when we write the `buf` below. Removing this inefficiency
+        // may not be worth the additional complexity.
+        self.blob.write_at(buf, offset).await?;
+
+        // Maintain the "buffer at tip" invariant by advancing offset to the end of this write if it
+        // extended the underlying blob.
+        buffer.offset = buffer.offset.max(end_offset);
+
         Ok(())
     }
 
     async fn truncate(&self, len: u64) -> Result<(), Error> {
-        // Acquire a write lock on the inner state
-        let mut inner = self.inner.write().await;
-
-        // Determine the current buffer boundaries
-        let buffer_start = inner.position;
-        let buffer_end = buffer_start + inner.buffer.len() as u64;
-
-        // Adjust buffer content based on truncation point
-        if len <= buffer_start {
-            // Truncation point is before or at the start of the buffer.
-            //
-            // All buffered data is now beyond the new length and should be discarded.
-            inner.buffer.clear();
-            inner.blob.truncate(len).await?;
-            inner.position = len;
-        } else if len < buffer_end {
-            // Truncation point is within the buffer.
-            //
-            // Keep only the portion of the buffer up to the truncation point.
-            let new_buffer_len = (len - buffer_start) as usize;
-            inner.buffer.truncate(new_buffer_len);
-        } else {
-            // Truncation point is at or after the end of the buffer.
-            //
-            // No changes needed to the buffer content.
+        // Acquire a write lock on the buffer.
+        let mut buffer = self.buffer.write().await;
+        if !buffer.truncate(len) {
+            self.blob.truncate(len).await?;
         }
+
         Ok(())
     }
 
     async fn sync(&self) -> Result<(), Error> {
-        let mut inner = self.inner.write().await;
-        inner.sync().await
+        let mut buffer = self.buffer.write().await;
+        if !buffer.is_empty() {
```

Comment:
I'm tempted to recommend just having `take_buf` return an `Option<>` rather than having to make multiple calls to buffer for these types of operations.